Feature/3575 make notes searchable

RK206 opened the pull request:
# Make notes searchable only for admin

journal and application indexes has to be re-indexed for this deployment

*Please don't delete any sections when completing this PR template; instead enter **N/A** for checkboxes or sections which are not applicable, unless otherwise stated below*

See https://github.com/DOAJ/doajPM/issues/3575

Describe the scope/purpose of the PR here in as much detail as you like

## Categorisation

This PR...
- [ ] has scripts to run
- [x] has migrations to run
- [ ] adds new infrastructure
- [ ] changes the CI pipeline
- [x] affects the public site
- [x] affects the editorial area
- [x] affects the publisher area
- [ ] affects the monitoring

## Basic PR Checklist

Instructions for developers:
* For each checklist item, if it is N/A to your PR check the N/A box
* For each item that you have done and confirmed for yourself, check Developer box (including if you have checked the N/A box)

Instructions for reviewers:
* For each checklist item that has been confirmed by the Developer, check the Reviewer box if you agree
* For multiple reviewers, feel free to add your own checkbox with your github username next to it if that helps with review tracking

### Code Style

- No deprecated methods are used
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer

- No magic strings/numbers - all strings are in `constants` or `messages` files
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer
  
- ES queries are wrapped in a Query object rather than inlined in the code
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer
  
- Where possible our common library functions have been used (e.g. dates manipulated via `dates`)
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer
  
- Cleaned up commented out code, etc
  - [ ] N/A
  - [x] Developer
  - [ ] Reviewer

### Testing

- Unit tests have been added/modified
  - [ ] N/A
  - [ ] Developer
  - [x] Reviewer
  
- Functional tests have been added/modified
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer
  
- Code has been run manually in development, and functional tests followed locally
  - [ ] N/A
  - [x] Developer
  - [ ] Reviewer

### Documentation

- FeatureMap annotations have been added
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer
  
- Documentation updates - if needed - have been identified and prepared for inclusion into main documentation (e.g. added and highlighted/commented as appropriate to this PR)
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer
  
- Core model documentation has been added to if needed: https://docs.google.com/spreadsheets/d/1lun2S9vwGbyfy3WjIjgXBm05D-3wWDZ4bp8xiIYfImM/edit
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer

- Events and consumers documentation has been added if needed: https://docs.google.com/spreadsheets/d/1oIeG5vg-blm2MZCE-7YhwulUlSz6TOUeY8jAftdP9JE/edit
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer
  
- The docs for this branch have been generated and pushed to the doc site (see docs/README.md for details)
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer


### Release Readiness

- If needed, migration has been created and tested locally
  - [ ] N/A
  - [x] Developer
  - [ ] Reviewer

- Release sheet has been created, and completed as far as is possible https://docs.google.com/spreadsheets/d/1Bqx23J1MwXzjrmAygbqlU3YHxN1Wf7zkkRv14eTVLZQ/edit#gid=2082243850
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer

- There has been a recent merge up from `develop` (or other base branch).  List the dates of the merges up from develop below
  - 2023-10-16
  - 2023-09-01
  - 2023-08-13


## Testing

List the Functional Tests that must be run to confirm this feature

(https://doaj.github.io/doaj-docs/feature/3575_make_notes_searchable/testbook/index.html#administrative_search/journals/test_notes_searchable_for_admin_and_not_for_public)



## Deployment

journal and application indexes has to be re-indexed for this deployment

### Configuration changes

N/A

### Scripts

N/A

### Migrations

Journal and application indexes has to be reindexed

### Monitoring

N/A

### New Infrastructure
N/A

### Continuous Integration

N/A

richard-jones commented on doajtest/testbook/administrative_search/journals.yml:
```diff
@@ -134,3 +134,22 @@ tests:
+      - The journal record is not displayed in the search results
+  - step: Search for the journal from api search with same words given in the note
+    results:
+      - The journal record is not displayed in the search results
```
Can we add here, or in the other functional tests, all the other cases that we want to make sure work:

1. That you *cannot* find the record via a keyword from the note in the public search
2. That you *can* find the record via a search in the admin searches (that's all contexts, admin, editor and associate editor)
3. That searching specifically on the "note" field will work in all those admin searches
4. That searching by a specific field other than "note" will not yield a result in all those admin searches

richard-jones commented on portality/settings.py:
```diff
@@ -484,7 +484,8 @@
             }
-        }
+        },
+        "copy_to": ["all_meta"]
```
Doing this on the type means that *all* objects which have unicode types now have a `copy_to` for `all_meta`, and I don't think we want that.  For example, this rule now appears in BackgroundJob records too.

We should look for an approach that adds this only when necessary, probably a type-specific override to these defaults for Journals and Applications

richard-jones commented on portality/settings.py:
```diff
@@ -681,7 +687,7 @@
             "role" : None,
             "query_validator" : "public_query_validator",
-            "query_filters" : ["only_in_doaj", "last_update_fallback"],
+            "query_filters" : ["only_in_doaj", "last_update_fallback", "search_all_meta"],
```
We also need to apply this constraint to the `journal,article` query route, otherwise that will allow you to search through notes via that endpoint (which is public).  Not sure on the implications of that, as articles do not have an `all_meta` field

richard-jones replied:
Hm, that could be a problem, because this would allow the notes data to "leak" through the fixed-query-widget's endpoint, so we need to think of a solution.  I don't really want to add an `all_meta` to articles unnecessarily, but it might be the only way.  Thoughts?

richard-jones commented on portality/settings.py:
```diff
@@ -713,15 +719,15 @@
             "auth" : True,
             "role" : "publisher",
-            "query_filters" : ["owner", "not_update_request"],
+            "query_filters" : ["owner", "not_update_request", "search_all_meta"],
```
Publisher `update_request` search also requires this constraint

richard-jones commented on portality/static/js/edges/admin.applications.edge.js:
```diff
@@ -61,7 +61,8 @@ $.extend(true, doaj, {
                         {'display':'Publisher','field':'bibjson.publisher.name'},
-                        {'display':'Journal: alternative title','field':'bibjson.alternative_title'}
+                        {'display':'Journal: alternative title','field':'bibjson.alternative_title'},
+                        {'display':'Notes','field':'admin.notes'},
```
Does this need to be `admin.notes.note`?

richard-jones commented on portality/static/js/edges/admin.update_requests.edge.js:
```diff
@@ -71,7 +71,8 @@ $.extend(true, doaj, {
                         {'display':'Publisher','field':'bibjson.publisher.name'},
-                        {'display':'Journal: Alternative Title','field':'bibjson.alternative_title'}
+                        {'display':'Journal: Alternative Title','field':'bibjson.alternative_title'},
+                        {'display':'Notes','field':'admin.notes'}
```
Needs to be `admin.notes.note`

RK206 commented:
All the issues have been addressed. Here is the summary.
- Modifying the url to query the notes from public page. This is rare case as the user should know the field name and notes to query. Even though user queries by updating the url, he will not see the notes as notes is not displayed in the public search results. This is fixed by adding a validator. Now it will return error message as the validator will fail the query.
- 'all_meta' has been moved to MAPPING_OPTS of journal and application and removed from shared struct
- Updated add_mapping_extensions method to add 'copy_to' for required mapping only
- Fixed all other issues mentioned

I have tested the feature locally by deleting the indexes and doing an anonymous import of the indexes journal and application.

richard-jones commented on doajtest/unit/test_query.py:
```diff
@@ -30,7 +31,12 @@
             "dao" : "portality.models.Journal"
-        }
+        },
+        "suggestion" : {
```
Can we call this `application` instead of `suggestion` - we're trying to slowly remove that terminology from the codebase.

RK206 replied:
I think we need to change at all places if we want to change suggestion to application as this is index_type. Let me know if I need to change at all places.

richard-jones replied:
You are quite right, leave as is for now

richard-jones commented on portality/settings.py:
```diff
@@ -900,6 +920,52 @@
 }
+# Exclude the fields that doesn't want to be searched by public queries
+# This is part of non_public_fields_validator.
+EXCLUDED_FIELDS = [
```
Can we give this a more specific name, like `PUBLIC_QUERY_VALIDATOR__EXCLUDED_FIELDS`, just so it's clear what it is out of context

RK206 replied:
updated

richard-jones commented on portality/settings.py:
```diff
@@ -626,6 +626,19 @@
+# This is useful when some extensions required for some objects additional to defaults.
+# ~~->DataObj:Library~~
+# ~~->Seamless:Library~~
+DATAOBJ_TO_MAPPING_EXTENSIONS = {
```
Can we give this a more descriptive name like `DATAOBJ_TO_MAPPING_COPY_TO_EXTENSIONS`

RK206 replied:
Updated

richard-jones commented on portality/settings.py:
```diff
@@ -900,6 +920,52 @@
+    "admin.notes.author_id"
+]
+
+ADMIN_NOTES_SEARCH_MAPPING = {
```
In the `settings.py` file can we group this with the other MAPPINGS keys, a bit above in the file

RK206 replied:
As per my understanding, above mappings are general mappings which apply for all. ADMIN_NOTES_SEARCH_MAPPING applies only for application and journal. So I have separated them. Please correct me if wrong.

richard-jones commented on portality/settings.py:
```diff
@@ -706,7 +719,7 @@
             "auth" : False,
             "role" : None,
-            "query_validator" : "public_query_validator",
+            "query_validators" : ["non_public_fields_validator", "public_query_validator"],
```
I'm happy to have two separate validators here chained together, OR for us to update the `public_query_validator` so that it can do both jobs.  If we're having 2 separate validators, maybe we should rename the `public_query_validator` to something more descriptive of what it does (which is to check paging parameters).

RK206 replied:
`non_public_fields_validator` is used at the places where `public_query_validator` not used. For example for Publisher application query (Line no 755) So I made them separate validators.

richard-jones reviewed (changes requested):
All looking good. A couple of very minor nitpicks about variable names, nothing of any great significance.  We do need to resolve the question of how we limit access to notes via the `journal,article` public endpoint, as I think at the moment it would be possible to query for notes via the route.  Possibly need to add this to our technical call next week.

richard-jones reviewed (approved):
I've now done a deep review of the security implications, and made some adjustments for absolute certainty, including one fix to how the mappings are generated.  I've also updated the functional tests a bit for ease of use.

From my perspective this is good to go
